src/devices/src/virtio/block/request.rs

```diff
@@ -18,28 +18,61 @@ use super::device::{CacheType, DiskProperties};
 use super::{Error, SECTOR_SHIFT, SECTOR_SIZE};
 
 #[derive(Debug)]
-pub enum ExecuteError {
+pub enum IoErrStatus {
     BadRequest(Error),
     Flush(io::Error),
-    Read(GuestMemoryError),
+    Read(u32, GuestMemoryError),
     Seek(io::Error),
     SyncAll(io::Error),
     Write(GuestMemoryError),
+}
+
+#[derive(Debug)]
+pub enum ErrStatus {
+    IoErr(IoErrStatus),
     Unsupported(u32),
 }
 
-impl ExecuteError {
-    pub fn status(&self) -> u32 {
-        match *self {
-            ExecuteError::BadRequest(_) => VIRTIO_BLK_S_IOERR,
-            ExecuteError::Flush(_) => VIRTIO_BLK_S_IOERR,
-            ExecuteError::Read(_) => VIRTIO_BLK_S_IOERR,
-            ExecuteError::Seek(_) => VIRTIO_BLK_S_IOERR,
-            ExecuteError::SyncAll(_) => VIRTIO_BLK_S_IOERR,
-            ExecuteError::Write(_) => VIRTIO_BLK_S_IOERR,
-            ExecuteError::Unsupported(_) => VIRTIO_BLK_S_UNSUPP,
+#[derive(Debug)]
+pub enum Status {
+    Ok(u32),
+    Err(ErrStatus),
+}
```

Comment:
nit: This looks like a glorified `Result<u32, ErrStatus>` with the upside that you can define your own methods on it. Is it worth it? Why not just process the contents of an actual `Result` or name the enum variants differently?

also nit: We could have a comment explaining that the `u32` is in fact `num_used_bytes`.

Reply:
I think it's worth it since the possibility to define our own methods improves the readability. I didn't see a performance penalty in my tests.

I added a comment.

Reply:
I commented this after reading `if let Status::Err(err_status) = status`, which was a bit confusing. When extracting an error, I expect something like `if let Err(e) = ...`. It's just a cosmetic thing at this point, so it's ok.